client/network/src/lib.rs

```diff
@@ -294,6 +296,12 @@ pub trait NetworkStateInfo {
 
 	/// Returns the local Peer ID.
 	fn local_peer_id(&self) -> PeerId;
+
+	/// Returns the local Peer PublicKey.
+	fn local_public_key(&self) -> PublicKey;
```

Comment:
Do I understand correctly that this pull request treats a `PublicKey` as an opaque data set? In other words, does this pull request make sure of `PublicKey` verification? If not, why not use the `PeerId` abstraction as the base network identity?

Reply:
There is no extra verification (except the length of the bytes)  for `PublicKey` in this PR, the reasons I didn't use `PeerId` are,
* `PeerId` is from libp2p, the encode/decode for it is pretty a challenge to me. I guess you mean using an opaque data like `Vec<u8>` to represent the PeerId when storing and convert it back when using it, that's probably also work in my scenario, but sounds a bit wild.
* It also needs to pass in and read by a user, I'm not sure how convenient to use it.
* trying to avoid let the runtime user to care about the network layer, and `PeerId` is something fall in this.

Reply:
> but sounds a bit wild

That's actually the preferred approach IMO.
The conversion from `PeerId` to `Vec<u8>` and vice versa is precisely spec'ced by libp2p and not something we have invented.

> It also needs to pass in and read by a user, I'm not sure how convenient to use it.
> trying to avoid the runtime user to care about the network layer, and PeerId is something fall in this.

In terms of difficulty don't really see the difference between letting users manipulate a `PublicKey` and letting users manipulate a `PeerId`.
It is true that if you actually want to analyse the content of the `PublicKey` or `PeerId`, then it's more complicated if you have a `PeerId`. However, I don't think users actually want to do that.
Furthermore, libp2p public keys are never exposed anywhere in the node. Users can easily see which `PeerId` they have, but not which public key they have.

Reply:
I was imagining a scenario that user generate node key so that it can be passed to the CLI when starting the node, at this time the public key is available and recorded by the user. Then be used in this node-authorization pallet.

I'm not in favor of putting PeerId to the user side. Even it's a mandatory concept in libp2p, but not so in Substrate runtime development AFAIK. Most of the time we are assuming the user don't need much knowledge such as PeerId about libp2p, isn't it?

Reply:
I just find this [P2P Networking](https://wiki.polkadot.network/docs/en/maintain-guides-how-to-setup-sentry-node#p2p-networking) in polkadot wiki, which eliminate a few concerns from me to use `PeerId` in runtime.
Still interested in others preference between PublicKey and PeerId.